go/geth_adapter/adapter.go

```diff
@@ -230,13 +194,215 @@ func (a *gethInterpreterAdapter) Run(contract *geth.Contract, input []byte, read
 	return res.Output, nil
 }
 
+func debugRunStart(input []byte, readOnly bool) {
+	fmt.Printf("Begin of interpreter:\n")
+	fmt.Printf("\tInput:  %v\n", input)
+	fmt.Printf("\tStatic: %v\n", readOnly)
+}
+
+func debugRunEnd(res tosca.Result) {
+	fmt.Printf("End of interpreter:\n")
+	fmt.Printf("\tSuccess:  %v\n", res.Success)
+	fmt.Printf("\tOutput:   %v\n", res.Output)
+	fmt.Printf("\tGas Left: %v\n", res.GasLeft)
+	fmt.Printf("\tRefund:   %v\n", res.GasRefund)
+}
+
+func shiftRefundBack(stateDB geth.StateDB, err error, refundShift uint64) {
+	if err == nil || err == geth.ErrExecutionReverted {
+		// In revert cases the accumulated refund to this point may be negative,
+		// which would cause the subtraction of the original refundShift to
+		// underflow the refund in the StateDB. Thus, the back-shift is capped
+		// by the available refund.
+		shift := refundShift
+		if cur := stateDB.GetRefund(); cur < shift {
+			shift = cur
+		}
+		stateDB.SubRefund(shift)
+	}
+}
+
+func removeReadOnlyInGas(depth int, readOnly bool, gas uint64) (bool, uint64) {
```

Comment:
nit: it would be nice to group the `setReadOnlyInGas` and `removeReadOnlyInGas` in the file such that they are next to each other;

Also, maybe consider renaming it to `encodeReadOnlyInGas` and `decodeReadOnlyFromGas` ...